File: frontend/src/types/rfp.ts
export enum RFPStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
  UNDER_REVIEW = 'under_review',
  CLOSED = 'closed',
  AWARDED = 'awarded',
  CANCELLED = 'cancelled'
}

export enum RFPType {
  REQUEST_FOR_PROPOSAL = 'rfp',
  REQUEST_FOR_QUOTATION = 'rfq',
  INVITATION_TO_BID = 'itb',
  REQUEST_FOR_INFORMATION = 'rfi'
}

export interface RFP {
  id: number
  title: string
  rfp_number: string
  description?: string
  rfp_type: RFPType
  status: RFPStatus
  estimated_budget?: number
  currency: string
  submission_deadline?: string
  evaluation_period?: number
  project_start_date?: string
  project_end_date?: string
  technical_requirements?: Record<string, any>
  evaluation_criteria?: Record<string, any>
  compliance_requirements?: Record<string, any>
  organization?: string
  contact_person?: string
  contact_email?: string
  contact_phone?: string
  issue_date?: string
  award_date?: string
  complexity_score?: number
  risk_assessment?: Record<string, any>
  go_no_go_recommendation?: string
  ai_analysis_summary?: string
  key_insights?: Record<string, any>
  created_by: number
  created_at: string
  updated_at: string
}

export interface RFPListItem {
  id: number
  title: string
  rfp_number: string
  status: RFPStatus
  rfp_type: RFPType
  estimated_budget?: number
  currency: string
  submission_deadline?: string
  organization?: string
  created_at: string
}

export interface CreateRFPData {
  title: string
  description?: string
  rfp_type: RFPType
  estimated_budget?: number
  currency?: string
  submission_deadline?: string
  evaluation_period?: number
  project_start_date?: string
  project_end_date?: string
  technical_requirements?: Record<string, any>
  evaluation_criteria?: Record<string, any>
  compliance_requirements?: Record<string, any>
  organization?: string
  contact_person?: string
  contact_email?: string
  contact_phone?: string
}

export interface UpdateRFPData extends Partial<CreateRFPData> {}

export interface RFPDocument {
  id: number
  filename: string
  original_filename: string
  file_size: number
  content_type: string
  document_type?: string
  processing_status: string
  created_at: string
}

export interface RFPAnalysisRequest {
  analyze_complexity?: boolean
  analyze_risks?: boolean
  generate_recommendations?: boolean
  custom_criteria?: Record<string, any>
}

export interface RFPAnalysisResponse {
  complexity_score?: number
  risk_assessment?: Record<string, any>
  go_no_go_recommendation?: string
  analysis_summary: string
  key_insights: Record<string, any>
  confidence_score?: number
  processing_time: number
}

export interface RFPFilters {
  status?: RFPStatus
  search?: string
  my_rfps?: boolean
  order_by?: string
  order_direction?: 'asc' | 'desc'
}

File: backend/models/rfp.py
"""
TenderWise AI - RFP Models
"""

from datetime import datetime
from typing import Optional
from sqlalchemy import Boolean, Column, DateTime, Integer, String, Text, JSON, ForeignKey, Enum, Numeric
from sqlalchemy.orm import relationship
from core.database import Base
import enum


class RFPStatus(str, enum.Enum):
    DRAFT = "draft"
    PUBLISHED = "published"
    UNDER_REVIEW = "under_review"
    CLOSED = "closed"
    AWARDED = "awarded"
    CANCELLED = "cancelled"


class RFPType(str, enum.Enum):
    REQUEST_FOR_PROPOSAL = "rfp"
    REQUEST_FOR_QUOTATION = "rfq"
    INVITATION_TO_BID = "itb"
    REQUEST_FOR_INFORMATION = "rfi"


class RFP(Base):
    """RFP (Request for Proposal) model"""
    
    __tablename__ = "rfps"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(500), nullable=False, index=True)
    rfp_number = Column(String(100), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    
    # RFP Details
    rfp_type = Column(Enum(RFPType), default=RFPType.REQUEST_FOR_PROPOSAL)
    status = Column(Enum(RFPStatus), default=RFPStatus.DRAFT, index=True)
    
    # Financial Information
    estimated_budget = Column(Numeric(15, 2), nullable=True)
    currency = Column(String(3), default="USD")
    
    # Timeline
    issue_date = Column(DateTime, nullable=True)
    submission_deadline = Column(DateTime, nullable=True)
    evaluation_period = Column(Integer, nullable=True)  # Days
    award_date = Column(DateTime, nullable=True)
    project_start_date = Column(DateTime, nullable=True)
    project_end_date = Column(DateTime, nullable=True)
    
    # Requirements and Criteria
    technical_requirements = Column(JSON, nullable=True)
    evaluation_criteria = Column(JSON, nullable=True)
    compliance_requirements = Column(JSON, nullable=True)
    
    # AI Analysis Results
    complexity_score = Column(Numeric(3, 2), nullable=True)  # 0.00 - 10.00
    risk_assessment = Column(JSON, nullable=True)
    go_no_go_recommendation = Column(String(50), nullable=True)
    ai_analysis_summary = Column(Text, nullable=True)
    key_insights = Column(JSON, nullable=True)
    
    # Metadata
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    organization = Column(String(255), nullable=True)
    contact_person = Column(String(255), nullable=True)
    contact_email = Column(String(255), nullable=True)
    contact_phone = Column(String(50), nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    # creator = relationship("User", back_populates="rfps")
    documents = relationship("RFPDocument", back_populates="rfp", cascade="all, delete-orphan")
    # proposals = relationship("Proposal", back_populates="rfp")

    def __repr__(self):
        return f"<RFP(id={self.id}, title='{self.title}', status='{self.status}')>"


class RFPDocument(Base):
    """RFP Document attachments"""
    
    __tablename__ = "rfp_documents"

    id = Column(Integer, primary_key=True, index=True)
    rfp_id = Column(Integer, ForeignKey("rfps.id"), nullable=False)
    
    # Document Information
    filename = Column(String(255), nullable=False)
    original_filename = Column(String(255), nullable=False)
    file_path = Column(String(500), nullable=False)
    file_size = Column(Integer, nullable=False)
    content_type = Column(String(100), nullable=False)
    
    # Document Processing
    extracted_text = Column(Text, nullable=True)
    document_type = Column(String(50), nullable=True)  # specifications, drawings, etc.
    processing_status = Column(String(50), default="pending")
    
    # Metadata
    uploaded_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    rfp = relationship("RFP", back_populates="documents")
    # uploader = relationship("User")

    def __repr__(self):
        return f"<RFPDocument(id={self.id}, filename='{self.filename}', rfp_id={self.rfp_id})>"